checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java

```diff
@@ -1,176 +1,873 @@
+import org.checkerframework.dataflow.cfg.node.AssignmentNode;
+import org.checkerframework.dataflow.cfg.node.LocalVariableNode;
 import org.checkerframework.dataflow.cfg.node.MethodInvocationNode;
 import org.checkerframework.dataflow.cfg.node.Node;
+import org.checkerframework.dataflow.cfg.node.ObjectCreationNode;
+import org.checkerframework.dataflow.expression.JavaExpression;
+import org.checkerframework.dataflow.expression.LocalVariable;
+import org.checkerframework.dataflow.util.NodeUtils;
 import org.checkerframework.javacutil.AnnotationBuilder;
+import org.checkerframework.javacutil.AnnotationUtils;
 import org.checkerframework.javacutil.BugInCF;
+import org.checkerframework.javacutil.TreePathUtil;
 import org.checkerframework.javacutil.TreeUtils;
+import org.checkerframework.javacutil.TypesUtils;
+import org.plumelib.util.CollectionsPlume;
 
 /**
- * This class contains the Resource Leak Checker's annotation inference algorithm. It contains
- * inference logic for owning annotations on final owning fields. It adds an @Owning annotation on a
- * field if it finds a method that satisfies the @MustCall obligation of the field along some path
- * to the regular exit point.
+ * This class implements the annotation inference algorithm for the Resource Leak Checker. It infers
+ * annotations such as {@code @}{@link Owning} on owning fields and
+ * parameters, @EnsuresCalledMethods on methods, and @InheritableMustCall on class declarations.
+ *
+ * <p>Each instance of this class corresponds to a single control flow graph (CFG), typically
+ * representing a method.
+ *
+ * <p>The algorithm determines if the @MustCall obligation of a field is satisfied along some path
+ * leading to the regular exit point of the method. If the obligation is satisfied, it adds
+ * an @Owning annotation on the field and an @EnsuresCalledMethods annotation on the method being
+ * analyzed. Additionally, if the method being analyzed satisfies the must-call obligation of all
+ * the enclosed owning fields, it adds a @InheritableMustCall annotation on the enclosing class.
+ *
+ * <p>Note: This class makes the assumption that the must-call set has only one element. Must-call
+ * sets with more than one element may be supported in the future.
+ *
+ * <p>See {@link ResourceLeakChecker#ENABLE_WPI_FOR_RLC} for an explanation of the meaning of the
+ * flags {@code -Ainfer} and {@code -AenableWpiForRlc}.
+ *
+ * @see <a
+ *     href="https://checkerframework.org/manual/#resource-leak-checker-inference-algo">Automatic
+ *     Inference of Resource Leak Specifications</a>
  */
 public class MustCallInference {
 
-  /** The set of owning fields. */
+  /**
+   * The fields that have been inferred to be disposed within the CFG currently under analysis. All
+   * of these fields will be given an @Owning annotation.
+   */
+  private final Set<VariableElement> disposedFields = new HashSet<>();
+
+  /** The fields that have been annotated with the {@code @Owning} annotation */
```

Comment:
Please avoid passive voice like "have been".  I don't know whether this means annotated by the programmer, or annotated by this algorithm so far, or something else.  Please clarify the documentation.